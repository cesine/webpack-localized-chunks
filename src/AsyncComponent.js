import React, { Component } from "react";
import queryString from 'query-string';

import locale from './locale';
const LOCALES = ['en', 'fr'];

export default function asyncComponent(importComponent) {
  class AsyncComponent extends Component {
    constructor(props) {
      super(props);

      window.addEventListener("routeChange", () => {
        this.renderLocalized();
      });

      this.state = {
        component: null
      };
    }

    renderLocalized = async() => {
      const { locale: activeLocale, clientSide } = queryString.parse(window.location.search);

      const sanitizedLocale = LOCALES.includes(activeLocale) ? activeLocale: 'fr';
      if (sanitizedLocale !== activeLocale) {
        console.warn(`invalid locale ${activeLocale} please try ${LOCALES}`);
      }
      const translations = await locale[sanitizedLocale]();

      const C = this.state.component;
      this.setState({
        App: C ? <C activeLocale={activeLocale} clientSide={clientSide!=='false'} translations={translations} {...this.props } /> : null
      });
    }

    async componentDidMount() {
      const { default: component } = await importComponent();

      this.setState({
        component: component
      });
      this.renderLocalized();
    }

    render() {
      return (
        <div>
          {this.state.App}
        </div>
      );
    }
  }

  return AsyncComponent;
}
