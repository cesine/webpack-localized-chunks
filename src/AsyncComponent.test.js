import React from 'react';
import ReactDOM from 'react-dom';
import AsyncComponent from './AsyncComponent';

describe('AsyncComponent', () => {
  it('should render empty until component is loaded', async () => {
    const AsyncApp = await AsyncComponent(() => import('./App'));
    const div = document.createElement('div');
    await ReactDOM.render(<AsyncApp clientSide={true} />, div);
    expect(div).toMatchSnapshot();
  });
});