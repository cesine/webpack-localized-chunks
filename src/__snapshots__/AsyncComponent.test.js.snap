// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncComponent should render empty until component is loaded 1`] = `
<div>
  <div />
</div>
`;
